### PR TITLE
Bookmarks

### PR DESCRIPTION
**Pull request by nancylee713**
This PR pertains to #11  

#### Attempt 1
```ruby
# I had to add two separate methods (i.e., `bookmarked_tutorials` and `bookmarked_videos`) in UserFacade to display tutorial name as well as bookmarked videos nested under a given tutorial. 

# This approach may be committing the "n+1 queries" problem, which should be avoided according to the user story. Is there a way to iterate over `user_videos` once without showing tutorial name at every iteration? I made this PR to ask for your input.
```

#### Attempt 2
```ruby
# We revamped the `#bookmarked_tutorials` method in UserFacade with ActiveRecord. 
# In order to use ActiveRecord queries, we had to add validations to all models (both missing and extended).
# Using `includes` method was key to avoiding n+1 query problem, as Fenton suggested.
# Still working on ordering bookmarked videos by position
```

#### Attempt 3 (working)
- Add an instance method (`#bookmarked_videos`) to User model, which still uses ActiveRecord associations
- Change the way bookmarked videos are displayed on user dashboard (aka, a simple link that contains both tutorial and video info)

### CODE REVIEW

**Comment by fentontaylor** [comment]
To avoid n + 1 queries, there is a method called `include` that also pulls in extra info during the first query. I forget the exact syntax though

File: app/facades/user_facade.rb [hunk 1]
```diff
@@ -3,6 +3,14 @@ def initialize(user)
...
   end
 
+  def bookmarked_tutorials
+    @user.user_videos.map { |uv| Tutorial.find(Video.find(uv.video_id).tutorial_id) }.uniq
+  end
+
+  def bookmarked_videos(tutorial)
+    @user.user_videos.map { |uv| Video.find(uv.video_id) }.select { |v| v.tutorial_id == tutorial.id }
+  end
+
```
Comment (fentontaylor) [inline app/facades/user_facade.rb]: Can these be written with ActiveRecord?

Comment (nancylee713) [reply app/facades/user_facade.rb]: You're right

File: app/facades/user_facade.rb [hunk 2]
```diff
@@ -3,6 +3,10 @@ def initialize(user)
     @user = user
   end
 
+  def bookmarked_tutorials
+    Tutorial.includes(:users).where(users: {id: @user.id})
```
Comment (nathangthomas) [inline app/facades/user_facade.rb]: Great use of .includes() here.

**Pull request merged**